test/Microsoft.AspNetCore.Http.Tests/StreamPipeWriterTests.cs

```diff
@@ -0,0 +1,498 @@
+
+            var task = Task.Run(async () =>
+            {
+                try
+                {
+                    var writingTask = pipeWriter.WriteAsync(Encoding.ASCII.GetBytes("data"));
+                    e.Set();
+                    flushResult = await writingTask;
+                }
+                catch (Exception ex)
+                {
+                    Console.WriteLine(ex.Message);
+                    throw ex;
+                }
+            });
+
+            e.Wait();
+
+            pipeWriter.CancelPendingFlush();
+
+            await task;
+
+            Assert.True(flushResult.IsCanceled);
+        }
+
+        [Fact]
+        public void FlushAsyncCancellationDeadlock()
+        {
+            var cts = new CancellationTokenSource();
+            var cts2 = new CancellationTokenSource();
+
+            PipeWriter buffer = Writer.WriteEmpty(MaximumSizeHigh);
+
+            var e = new ManualResetEventSlim();
+
+            ValueTaskAwaiter<FlushResult> awaiter = buffer.FlushAsync(cts.Token).GetAwaiter();
+            awaiter.OnCompleted(
+                () => {
+                    // We are on cancellation thread and need to wait until another FlushAsync call
+                    // takes pipe state lock
+                    e.Wait();
+
+                    // Make sure we had enough time to reach _cancellationTokenRegistration.Dispose
+                    Thread.Sleep(100);
+
+                    // Try to take pipe state lock
+                    buffer.FlushAsync();
+                });
+
+            // Start a thread that would run cancellation callbacks
+            Task cancellationTask = Task.Run(() => cts.Cancel());
+            // Start a thread that would call FlushAsync with different token
+            // and block on _cancellationTokenRegistration.Dispose
+            Task blockingTask = Task.Run(
+                () => {
+                    e.Set();
+                    buffer.FlushAsync(cts2.Token);
+                });
+
+            bool completed = Task.WhenAll(cancellationTask, blockingTask).Wait(TimeSpan.FromSeconds(10));
+            Assert.True(completed);
+        }
+
+        [Fact]
+        public void FlushAsyncCompletedAfterPreCancellation()
+        {
+            PipeWriter writableBuffer = Writer.WriteEmpty(1);
+
+            Writer.CancelPendingFlush();
+
+            ValueTask<FlushResult> flushAsync = writableBuffer.FlushAsync();
+
+            Assert.True(flushAsync.IsCompleted);
+
+            FlushResult flushResult = flushAsync.GetAwaiter().GetResult();
+
+            Assert.True(flushResult.IsCanceled);
+
+            flushAsync = writableBuffer.FlushAsync();
+
+            Assert.True(flushAsync.IsCompleted);
+        }
+
+        [Fact]
+        public void FlushAsyncReturnsCanceledIfCanceledBeforeFlush()
+        {
+            CheckCanceledFlush();
+        }
+
+        [Fact]
+        public void FlushAsyncReturnsCanceledIfCanceledBeforeFlushMultipleTimes()
```

Comment:
Don't mix tests copied from Pipes and new ones. It would be very hard to merge them back into corefx.